### PR TITLE
Audits

### LINKED ISSUE #36: Write flow describing the goals of the audit log
This should define the user flow for retrieving an audit log for a user or key.

- [ ] Write out the flow
- [ ] Define the properties we want out of the log (e.g. lists all attempted actions, defines responsible parties if an action failed, always accessible as long as the key server is accessible, etc.).

### PR DESCRIPTION
**Pull request by indomitableSwan**
Closes #36.

Would like to request @vidyaap to collaborate on finishing this draft PR:
- Add any technical details/clarifications that seem appropriate.
- Mark it as ready for review.

Anyone on the team should be able to review the completed PR.

### CODE REVIEW

File: cryptographic_flows.md [hunk 1]
```diff
@@ -22,8 +22,36 @@ An asset owner that has not previously interacted with the key server MUST regis
...
+## Retrieve Audit Logs
+The asset owner can request audit logs from the key server.
+
+Input:
+- `user_id`, a 128-bit globally unique identifier (GUID) representing the identity of the asset owner.
+- `type`, one of:
+    - `"system only"`, which indicates the asset owner wants a record of registration and login actions;
+    - `"key only"`, which indicates the asset owner wants a record of requested key use operations with respect to or more keys; or
+    - `"all"`, which indicates the asset owner wants both system and requested key use operations.
+- `key_identifiers`, an OPTIONAL list of key identifiers.
```
Comment (vidyaap) [inline cryptographic_flows.md:37]: If there is no `key identifier` provided but the `type` is `“key only”` is the output the logs for all the keys?

File: cryptographic_flows.md [hunk 2]
```diff
@@ -22,8 +22,36 @@ An asset owner that has not previously interacted with the key server MUST regis
...
 1. The client and server close the session upon completion of the request.
 
+## Retrieve Audit Logs
+The asset owner can request audit logs from the key server.
+
+Input:
+- `user_id`, a 128-bit globally unique identifier (GUID) representing the identity of the asset owner.
+- `type`, one of:
+    - `"system only"`, which indicates the asset owner wants a record of registration and login actions;
+    - `"key only"`, which indicates the asset owner wants a record of requested key use operations with respect to or more keys; or
```
Comment (vidyaap) [inline cryptographic_flows.md:35]: Is this meant to be with respect to “one or more” keys?

File: cryptographic_flows.md [hunk 3]
```diff
@@ -22,8 +22,36 @@ An asset owner that has not previously interacted with the key server MUST regis
     1. Checks that `user_id` matches that of the authenticated user in the open registration session. If this check fails, the server MUST abort.
     1. Checks that the received ciphertext is of the expected format and length.
     1. [Stores](#server-side-storage) the received ciphertext in a record associated with `user_id`.
+    1. Stores the registration registration request in the [audit log](#audit-logs).
 1. The client and server close the session upon completion of the request.
 
+## Retrieve Audit Logs
+The asset owner can request audit logs from the key server.
+
+Input:
```
Comment (vidyaap) [inline cryptographic_flows.md:31]: Can we include an optional date range as a parameter?

File: cryptographic_flows.md [hunk 4]
```diff
@@ -22,8 +22,36 @@ An asset owner that has not previously interacted with the key server MUST regis
...
+
+Protocol:
+
+1. The client:
+   1. [Opens a request session](systems-architecture.md#request-session) for the given user id `user_id`. 
+   1. Sends a request message to the key server over the session's secure channel. This message MUST indicate the desire to retrieve audit logs, specify the type of audit requested (i.e., `"system only"`, `"key only"`, or `"all"`) and contain `user_id`.
+1. The key server:
+   1. Runs a validity check on the received request and `user_id`(i.e., there must be a valid open request session, the request must conform to the expected format, and `user_id` must be of the expected format and length, and should match that of the open request session). If this check fails, the server MUST reject the request.
+   1. Retrieves all log records relevant to the client's request, summarises these records in `summary_record`, and sends `summary_record` to the client over the secure channel.
+   1. Stores the current request information, including the outcome of the validity check, in an [audit log](#audit-logs) associated with the given user.
```
Comment (vidyaap) [inline cryptographic_flows.md:49]: Does this mean we make a log entry for someone having requested the logs? Also does this type of log fall under "system log"?

File: cryptographic_flows.md (hunk 1 above)
Comment (indomitableSwan) [reply cryptographic_flows.md:37]: That would make sense. It should probably do the same for "all" with no key identifier provided, right?

File: cryptographic_flows.md (hunk 4 above)
Comment (indomitableSwan) [reply cryptographic_flows.md:49]: Yes, and yes.

File: cryptographic_flows.md (hunk 1 above)
Comment (vidyaap) [reply cryptographic_flows.md]: Yes I think that makes sense. Added a bit about that in the `key_identifiers` description.

File: cryptographic_flows.md (hunk 3 above)
Comment (jakinyele) [reply cryptographic_flows.md:31]: Could we consider adding severity levels across log entry types? Will provide additional granularity and implicit log filtering on requests. Some common levels: `error`, `warning`, `debug`, `info`, and `notice`.

File: cryptographic_flows.md [hunk 5]
```diff
@@ -164,6 +245,8 @@ For now, simple clear-text storage is acceptable.
 
 #### Server-side storage
 - [TODO #28](https://github.com/boltlabs-inc/key-mgmt-spec/issues/28). Include appropriate requirements for server-side secure storage and generate relevant issues in key-mgmt.
+##### Audit Logs
+The server storage should include a per-user audit log that tracks system registration and logins, key use requests, and audit log requests. A single log entry contains the following information about each action: action, actor, date, outcome (e.g. success or failure), any related key identifier.
```
Comment (jakinyele) [inline cryptographic_flows.md:249]: Is this the right place to discuss requirements around log rotation and backup for per-user audit logs? 

This is largely a deployment issue but would like to figure out the right balance between the availability of logs and preventing DoS on server storage.

**Review by jakinyele — APPROVED** [review]
Looks good but left a few comments about severity levels and log rotation/backup of log entries.

**Pull request merged**